### ISSUE TITLE
Add GetAlbumPhotos overload method in PhotoStationPersonalService to obtain all photos in album with no default limit

### ISSUE THREAD

**Issue by let92** [Arceden/SynologyNet#2]
Hi, is it possible to add an overloaded method for **GetAlbumPhotos** in **PhotoStationPersonalService** class in order to obtain the list of photos in the album without the limit set to 100 by default?

I hope for a positive response and thanks anyway for your availability.

**Comment by Arceden** [comment]
Hey @let92, 
I believe the Synology endpoint does require the user to set a limit. But i suppose it would make sense for this limit to be as high as possible and letting the user decide to lower it if needed. Which also reminds me that i forgot to add the [CollectionFilter](https://github.com/Arceden/SynologyNet/blob/master/SynologyNet/Models/Requests/Photo/CollectionFilter.cs) model as parameter so that you can change any filter.

I will try to add the filter parameter to the method soon

**Comment by let92** [comment]
> Hey @let92, I believe the Synology endpoint does require the user to set a limit. But i suppose it would make sense for this limit to be as high as possible and letting the user decide to lower it if needed. Which also reminds me that i forgot to add the [CollectionFilter](https://github.com/Arceden/SynologyNet/blob/master/SynologyNet/Models/Requests/Photo/CollectionFilter.cs) model as parameter so that you can change any filter.
> 
> I will try to add the filter parameter to the method soon

Understood, thank you @Arceden.

I've found another problem, tell me if I have to move it to another issue: the **Passphrase** property of the **Album** class is not evaluated (it is set to "") when called **GetAlbums** and when calling **GetAlbumPhotos** I've noticed that the method returns also photos from others albums.

Do you encounter the same problem?

Thank you again for your availability @Arceden.

**Comment by Arceden** [comment]
I have expanded my personal testing environment and seem to be running into the same issue now.

**Comment by Arceden** [comment]
The bug within GetAlbumPhotos has been fixed in the new corresponding branch and will be merged into master once i figure out a good way to handle filtering options. 👍

**Comment by let92** [comment]
> The bug within GetAlbumPhotos has been fixed in the new corresponding branch and will be merged into master once i figure out a good way to handle filtering options. 👍

Do you intend the bug within **GetAlbums**?
If so, can't you just publish the fix so I can do more testing for my use case of the library? 😃

I've managed to solve the problem regarding the default limit set in **GetAlbumPhotos** creating a class child of **PersonalPhotoRepository** and adding a method that calls **GetAlbumPhotos** with custom limits set by me.

Thanks again for the patience and availability 😊

**Comment by Arceden** [comment]
> Do you intend the bug within GetAlbums?
If so, can't you just publish the fix so I can do more testing for my use case of the library? 😃

Good suggestion, i will roll out a patch so you can properly use the GetAlbums() call earlier

**Comment by let92** [comment]
> > Do you intend the bug within GetAlbums?
> > If so, can't you just publish the fix so I can do more testing for my use case of the library? 😃
> 
> Good suggestion, i will roll out a patch so you can properly use the GetAlbums() call earlier

Thank you very much! I'll update the package and try again in the next days, I'll keep you updated.

Thank you again for all!

**Comment by let92** [comment]
Hi @Arceden, also with version **0.1.1** of the package the **Passphrase** property of the **Album** class is set to "" when called **GetAlbums** and calling **GetAlbumPhotos** returns no photos even with your recent commit.

However, I noticed that **GetSharedAlbums** evaluates the **Passphrase** property.

Do you encounter the same problem?

Let me know and thanks again for the patience.

**Comment by let92** [comment]
Hi @Arceden, also with version **0.1.1** of the package the **Passphrase** property of the **Album** class is set to "" when called **GetAlbums** and calling **GetAlbumPhotos** returns no photos even with your recent commit.

However, I noticed that **GetSharedAlbums** evaluates the **Passphrase** property.

The difference I see between the two methods is the **Version** attribute setted in **GetSharedAlbums** method that is missing in **GetAlbums** method.

Maybe is this the problem?

Let me know what you think and thanks again for the patience.

**Comment by Arceden** [comment]
Hey @let92, i think i misunderstood your question about the passphrase the first time. 

The **Passphrase** field is only being returned by the Synology Api when the album is shared with the current user. This is because the current user is not the owner of the shared album and needs a passphrase to get acccess to the contents of the album. This is not required when the album is owned by the current user.

So in short;
- Synology does not return a passphrase for the GetAlbums() endpoint
- Synology does return a passphrase for the GetSharedAlbums() endpoint

I believe there is no bug happening here, but more something that happens by design 👍 
Hope this clears up some confusion

**Comment by Arceden** [comment]
If it helps, i found someone on GitHub who made some documentation for the Synology PhotoStation Api which also describes the required parameters for albums.

https://github.com/zeichensatz/SynologyPhotosAPI/blob/master/API/SYNO.Foto/SYNO.Foto.Browse.Item.list.md

**Comment by let92** [comment]
> Hey @let92, i think i misunderstood your question about the passphrase the first time.
> 
> The **Passphrase** field is only being returned by the Synology Api when the album is shared with the current user. This is because the current user is not the owner of the shared album and needs a passphrase to get acccess to the contents of the album. This is not required when the album is owned by the current user.
> 
> So in short;
> 
> * Synology does not return a passphrase for the GetAlbums() endpoint
> * Synology does return a passphrase for the GetSharedAlbums() endpoint
> 
> I believe there is no bug happening here, but more something that happens by design 👍 Hope this clears up some confusion

Clearly understood now, but with the current implementation are you able to retrieve the photos in an album?

I can't make it 😓

**Comment by Arceden** [comment]
You should be able to retrieve photos from a normal and shared album with v0.1.1. What kind of error are you running up against? (exceptions or unexpected results)

In order to fetch photos from an album, you would need to do the following steps:
- Authenticate
- GetAlbums or GetSharedAlbums
- GetAlbumPhotos or GetSharedAlbumPhotos
- DownloadPhoto to get file itself as bytearray

I have some example code in the SynologyConsole Program.cs file which might be helpful. But i will probably make make some more easly readable documentation soon aswell

**Comment by let92** [comment]
This is the code I use.

```
_synologyConnection = new Synology(Settings.Default.SynologyAddress, Settings.Default.SynologyUser, Settings.Default.SynologyPassword);
await _synologyConnection.Authentication.Login();

IPhotoStationService photoStation = _synologyConnection.PhotoStation;
var personal = photoStation.Personal;
var albums = await personal.GetAlbums();

foreach (var album in albums)
{
	var albumPhotos = await personal.GetAlbumPhotos(album);
	...
}
```

The authentication is fine and **GetAlbums** returns the entire list of my Synology Photos albums.

To note the **ItemCount** of each **Album** entity is correct (image 1), but when calling **GetAlbumPhotos** I got only 2 photos for each album that are the same for each album and I don't know which photos are since the FileName is null (image 2).

![Arceden_SynologyNET_Problem with retrieving album photos_1](https://user-images.githubusercontent.com/51446253/211299664-4a045d05-aa0c-477e-8b7e-8db22240aed9.jpg)

![Arceden_SynologyNET_Problem with retrieving album photos_2](https://user-images.githubusercontent.com/51446253/211299759-11990bba-2dd6-4b09-b92c-3aa340368a35.jpg)

Furthermore, I noticed that the API **SYNO.Foto.Browse.Item** has two versions (image 3), but the method **GetAlbumPhotos** hasn't the **Version** attribute.

![Arceden_SynologyNET_Problem with retrieving album photos_3](https://user-images.githubusercontent.com/51446253/211299852-8a1161e6-9b53-447f-8272-b21b41ab3875.jpg)

Do you have the same situation?

I hope you can help me to solve the problem.

Thanks again for all.

**Comment by Arceden** [comment]
The Synology repositories in this codebase automatically target version 1 of an endpoint if no version attribute was set manually. Perhaps it could somehow be related, but i seem to be having issues reproducing the bug. 
- What limit did you choose for your GetAlbumPhotos call? Synology has a limit of about 5000.
- What is your version of DSM? This might affect the way the API responds to calls. I am running on DSM 7.1.1.
- Do you have any special types of files on your PhotoStation album?

Hope we can find a fix for this odd issue

**Comment by let92** [comment]
- I set 300 as limit, but also with 100 the problem persists;
- I own a DS920+ with DSM 7.1.1 like you;
- some albums has some videos.

Anyway, I've forked your repository for debugging the project and I think I've found the problem:
the result of **BaseRepository.GetMethod** for **PrepareRequest** in **PersonalPhotoRepository.GetAlbumPhotos** method returns the API "**SYNO.Foto.Sharing.Misc**" and not "**SYNO.Foto.Browse.Item**".

![Arceden_SynologyNET_Problem with retrieving album photos_4](https://user-images.githubusercontent.com/51446253/212023356-e77ab5f0-1e8e-47a1-a692-b77c3bcadf7a.jpg)

But I've never called the method **GetSharedAlbums**, as you can see in the call stack.

![Arceden_SynologyNET_Problem with retrieving album photos_5](https://user-images.githubusercontent.com/51446253/212023079-c9bec816-cc96-4e72-a611-1177333ff675.jpg)

Don't you encounter the same problem?

**Comment by Arceden** [comment]
I have not run into the same problem, but thank you for showing these debugging images. I think i know a way i could fix this issue

**Comment by let92** [comment]
Thanks to you, it's really a strange situation.

Hope you can solve the problem!

**Comment by Arceden** [comment]
I have released a new version of the library which has a default limit of 5000 and also allows you to set the pagination filtering yourself through the optional parameters. Example;
`await photoStation.GetSharedAlbums(pagingFilter: new() { Limit = 250 });`

I believe your bug might be related the overloaded/overridden method you made to set the limit to something higher than 100 which confuses the logic i had to interpret the Synology Request attributes within the repository. While this new version of the SynologyNet library might also prevent the bug from happening, i will also still try to improve the attribute logic to prevent simelar issues from happening.

If the issue still happens while using the new version, let me know 😄

**Comment by let92** [comment]
Very good update! Now I am able to see correctly every element in every album (also videos, thanks!)
Thank you very much!

Maybe do you have a guide or list with methods and parameters  for all Synology Photos API available? 

I mean, in the documentation present at the below link (the one you linked above) there are some APIs listed, but not all documented, like creation/deletion of albums and adding photo to albums.

https://github.com/zeichensatz/SynologyPhotosAPI

Which are the parameters to send to the API? How did you come to know what to send to the APIs used by your library?

My goal is to create a .NET application for creating albums in Synology Photos based on the albums present in my Google Photos account, and then adding also the photos in Google Photos albums (already present on the NAS) to the new Synology Photos albums.

If I manage to develop the project, at the end I'll contribute to your project adding the APIs used by my application (at least creation of albums and album photo adding).

I hope you can tell me how to know what to send to the Synology APIs, thanks again for all!

**Comment by Arceden** [comment]
Great! Good to see that the issue has been resolved. 

My only references for this project are that GitHub repository by zeichensatz and the Synology Developer guide. But i unfortunally cannot find official documentation for the PhotoStation. Perhaps that would be released later. https://www.synology.com/en-global/support/developer#tool

My other option is to use the GetInfo() list to see all available endpoint and attempt to work from there, which is a bit difficult.

**Comment by let92** [comment]
Understood, thanks again for all!
